File: src/main/kotlin/io/eontimer/service/CalibrationService.kt
package io.eontimer.service

import io.eontimer.timer.Settings
import org.springframework.stereotype.Service
import kotlin.math.roundToLong

@Service
class CalibrationService(
    private val timerSettings: Settings
) {
    private val console get() = timerSettings.console.get()

    fun toMillis(delays: Long): Long =
        (delays * console.frameRate).roundToLong()

    fun toDelays(millis: Long): Long =
        (millis / console.frameRate).roundToLong()

    fun createCalibration(delay: Long, second: Long): Long =
        toMillis(delay - toDelays(second * 1000))

    fun calibrateToMillis(value: Long): Long {
        return if (timerSettings.precisionCalibrationMode.get()) value else toMillis(value)
    }

    fun calibrateToDelays(value: Long): Long {
        return if (timerSettings.precisionCalibrationMode.get()) value else toDelays(value)
    }
}

File: src/Util/CalibrationService.cpp
//
// Created by Dylan Meadows on 2020-03-10.
//

#include "CalibrationService.h"
#include <cmath>

namespace EonTimer::Util {
    CalibrationService::CalibrationService(const Timer::Settings *timerSettings) : timerSettings(timerSettings) {}

    i32 CalibrationService::toDelays(const std::chrono::milliseconds ms) const {
        const double framerate = Timer::getFrameRate(timerSettings->getConsole());
        return static_cast<i32>(std::round(ms.count() / framerate));
    }

    std::chrono::milliseconds CalibrationService::toMilliseconds(const i32 delays) const {
        const double framerate = Timer::getFrameRate(timerSettings->getConsole());
        return std::chrono::milliseconds(static_cast<i32>(std::round(delays * framerate)));
    }

    i32 CalibrationService::calibrateToDelays(const std::chrono::milliseconds ms) const {
        return timerSettings->isPrecisionCalibrationEnabled() ? ms.count() : toDelays(ms);
    }

    std::chrono::milliseconds CalibrationService::calibrateToMilliseconds(i32 delays) const {
        return timerSettings->isPrecisionCalibrationEnabled() ? std::chrono::milliseconds(delays)
                                                              : toMilliseconds(delays);
    }

    i32 CalibrationService::createCalibration(const i32 delays, const i32 seconds) const {
        return toMilliseconds(delays - toDelays(std::chrono::milliseconds(seconds * 1000))).count();
    }
}  // namespace EonTimer::Util